src/Workspaces/Remote/Core/ExportProviderBuilder.cs

```diff
@@ -0,0 +1,198 @@
+            assemblyPaths = assemblyPaths.Sort(StringComparer.Ordinal);
+
+            var hashContents = new StringBuilder();
+
+            // This should vary based on .NET runtime major version so that as some of our processes switch between our target
+            // .NET version and the user's selected SDK runtime version (which may be newer), the MEF cache is kept isolated.
+            // This can be important when the MEF catalog records full assembly names such as "System.Runtime, 8.0.0.0" yet
+            // we might be running on .NET 7 or .NET 8, depending on the particular session and user settings.
+            hashContents.Append(Environment.Version.Major);
+
+            foreach (var assemblyPath in assemblyPaths)
+            {
+                // Include assembly path in the hash so that changes to the set of included
+                // assemblies cause the composition to be rebuilt.
+                hashContents.Append(assemblyPath);
+                // Include the last write time in the hash so that newer assemblies written
+                // to the same location cause the composition to be rebuilt.
+                hashContents.Append(File.GetLastWriteTimeUtc(assemblyPath).ToString("F"));
+            }
+
+            // Create base64 string of the hash.
+            var hashAsBase64String = Checksum.Create(hashContents.ToString()).ToString();
+
+            // Convert to filename safe base64 string.
+            return hashAsBase64String.Replace('+', '-').Replace('/', '_').TrimEnd('=');
+        }
+    }
+
+    protected virtual async Task WriteCompositionCacheAsync(string compositionCacheFile, CompositionConfiguration config, CancellationToken cancellationToken)
+    {
+        try
+        {
+            await Task.Yield().ConfigureAwait(false);
+
+            var directory = Path.GetDirectoryName(compositionCacheFile)!;
+            var directoryInfo = Directory.CreateDirectory(directory);
+            PerformCacheDirectoryCleanup(directoryInfo, cancellationToken);
+
+            CachedComposition cachedComposition = new();
+            var tempFilePath = Path.Combine(directory, Path.GetRandomFileName());
+            using (FileStream cacheStream = new(tempFilePath, FileMode.Create, FileAccess.Write, FileShare.None, bufferSize: 4096, useAsync: true))
+            {
+                await cachedComposition.SaveAsync(config, cacheStream, cancellationToken).ConfigureAwait(false);
+            }
+
+            File.Move(tempFilePath, compositionCacheFile);
+        }
+        catch (Exception ex)
+        {
+            LogError($"Failed to save MEF cache: {ex}");
+        }
+    }
+
+    protected virtual void PerformCacheDirectoryCleanup(DirectoryInfo directoryInfo, CancellationToken cancellationToken)
+    {
+        // Delete any existing cached files.
+        foreach (var fileInfo in directoryInfo.EnumerateFiles())
+        {
+            IOUtilities.PerformIO(fileInfo.Delete);
```

Comment:
doc that you're itnentionally making it so that failing to delete isn't an issue.